internal/generate/collector/filter.go

```diff
@@ -0,0 +1,177 @@
+		}
+	}
+	c.CustomResources = customResources
+}
+
+// deduplicate removes duplicate objects from the collection, since we are
+// collecting an arbitrary list of manifests.
+func (c *Manifests) deduplicate() error {
+	hashes := make(map[string]struct{})
+
+	roles := []rbacv1.Role{}
+	for _, role := range c.Roles {
+		hasHash, err := addToHashes(&role, hashes)
+		if err != nil {
+			return err
+		}
+		if !hasHash {
+			roles = append(roles, role)
+		}
+	}
+	c.Roles = roles
+
+	clusterRoles := []rbacv1.ClusterRole{}
+	for _, clusterRole := range c.ClusterRoles {
+		hasHash, err := addToHashes(&clusterRole, hashes)
+		if err != nil {
+			return err
+		}
+		if !hasHash {
+			clusterRoles = append(clusterRoles, clusterRole)
+		}
+	}
+	c.ClusterRoles = clusterRoles
+
+	deps := []appsv1.Deployment{}
+	for _, dep := range c.Deployments {
+		hasHash, err := addToHashes(&dep, hashes)
+		if err != nil {
+			return err
+		}
+		if !hasHash {
+			deps = append(deps, dep)
+		}
+	}
+	c.Deployments = deps
+
+	crds := []apiextv1beta1.CustomResourceDefinition{}
+	for _, crd := range c.CustomResourceDefinitions {
+		hasHash, err := addToHashes(&crd, hashes)
+		if err != nil {
+			return err
+		}
+		if !hasHash {
+			crds = append(crds, crd)
+		}
+	}
+	c.CustomResourceDefinitions = crds
+
+	validatingWebhooks := []admissionregv1.ValidatingWebhook{}
+	for _, webhook := range c.ValidatingWebhooks {
+		hasHash, err := addToHashes(&webhook, hashes)
+		if err != nil {
+			return err
+		}
+		if !hasHash {
+			validatingWebhooks = append(validatingWebhooks, webhook)
+		}
+	}
+	c.ValidatingWebhooks = validatingWebhooks
+
+	mutatingWebhooks := []admissionregv1.MutatingWebhook{}
+	for _, webhook := range c.MutatingWebhooks {
+		hasHash, err := addToHashes(&webhook, hashes)
+		if err != nil {
+			return err
+		}
+		if !hasHash {
+			mutatingWebhooks = append(mutatingWebhooks, webhook)
+		}
+	}
+	c.MutatingWebhooks = mutatingWebhooks
+
+	crs := []unstructured.Unstructured{}
+	for _, cr := range c.CustomResources {
+		b, err := cr.MarshalJSON()
+		if err != nil {
+			return err
+		}
+		hash := hashContents(b)
+		if _, hasHash := hashes[hash]; !hasHash {
+			crs = append(crs, cr)
+			hashes[hash] = struct{}{}
+		}
+	}
+	c.CustomResources = crs
+
+	return nil
+}
+
+// marshaller is an interface used to generalize hashing for deduplication.
+type marshaller interface {
+	Marshal() ([]byte, error)
+}
+
+// addToHashes calls m.Marshal(), hashes the returned bytes, and adds the
+// hash to hashes if it does not exist. addToHashes returns true if m's hash
+// was not in hashes.
+func addToHashes(m marshaller, hashes map[string]struct{}) (bool, error) {
+	b, err := m.Marshal()
+	if err != nil {
+		return false, err
+	}
+	hash := hashContents(b)
+	_, hasHash := hashes[hash]
+	if !hasHash {
+		hashes[hash] = struct{}{}
+	}
+	return hasHash, nil
+}
+
+// hashContents creates a sha256 md5 digest of b's bytes.
+func hashContents(b []byte) string {
+	h := sha256.New()
+	_, _ = h.Write(b)
```

Comment:
Nit: I realize this is just copying over but maybe we should log the error at the very least.
So perhaps a TODO for this.

Reply:
This will never return an error (looking at the implementation) but if std changes it may. I will follow up with an error check.